apm-sniffer/apm-sdk-plugin/undertow-plugins/undertow-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/undertow/v2x/handler/TracingHandler.java

```diff
@@ -0,0 +1,88 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.apache.skywalking.apm.plugin.undertow.v2x.handler;
+
+import io.undertow.server.ExchangeCompletionListener;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.HeaderMap;
+import org.apache.skywalking.apm.agent.core.context.CarrierItem;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+import org.apache.skywalking.apm.plugin.undertow.v2x.Constants;
+
+/**
+ * @author AI
+ * 2019-08-06
+ */
+public class TracingHandler implements HttpHandler {
+    private final String template;
+    private final HttpHandler next;
+
+    public TracingHandler(HttpHandler handler) {
+        this(null, handler);
+    }
+
+
+    public TracingHandler(String template, HttpHandler handler) {
+        this.next = handler;
+        this.template = template;
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        final HeaderMap headers = exchange.getRequestHeaders();
+        final ContextCarrier carrier = new ContextCarrier();
+        CarrierItem items = carrier.items();
+        while (items.hasNext()) {
+            items = items.next();
+            items.setHeadValue(headers.getFirst(items.getHeadKey()));
+        }
+        String operationName;
+        if (null == template) {
+            operationName = exchange.getRequestPath();
+        } else {
+            operationName = template;
+        }
+        final AbstractSpan span = ContextManager.createEntrySpan(operationName, carrier);
+        Tags.URL.set(span, exchange.getRequestURL());
+        Tags.HTTP.METHOD.set(span, exchange.getRequestMethod().toString());
+        span.setComponent(ComponentsDefine.UNDERTOW);
+        SpanLayer.asHttp(span);
+        span.prepareForAsync();
+        exchange.addExchangeCompleteListener(new ExchangeCompletionListener() {
+            @Override
+            public void exchangeEvent(HttpServerExchange httpServerExchange, NextListener nextListener) {
+                if (httpServerExchange.getStatusCode() >= 400) {
+                    span.errorOccurred();
+                    Tags.STATUS_CODE.set(span, Integer.toString(httpServerExchange.getStatusCode()));
+                }
+                span.asyncFinish();
+                nextListener.proceed();
+            }
+        });
+        next.handleRequest(exchange);
```

Comment:
We cannot guarantee that there is no bug in the plugin, and we try our best to avoid effecting the target application, that's why we have so many `try-catch`s in the core of the agent,

https://github.com/apache/skywalking/blob/5e8410834de20254c2bdab2d86cdbfd79d600904/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/interceptor/enhance/InstMethodsInter.java#L81-L111

But there is no protection here and if `span.prepareForAsync();` throws (introduced in #3201 ), the handler chain breaks and the handlers of the target application won't work (`next.handleRequest` won't be called)

Reply:
I think this exception should be a development stage protection, don't need runtime check. Do you agree? Or do you have any scenario could make this undetected in dev stage?

Reply:
That's what I'm not sure about,  maybe I overreact on the exception :)